darts/utils/optimised_historical_forecasts.py

```diff
@@ -0,0 +1,427 @@
+                times=times[0]
+                if stride == 1 and model.output_chunk_length == 1
+                else generate_index(
+                    start=hist_fct_start + (forecast_horizon - 1) * freq,
+                    length=forecast.shape[0],
+                    freq=freq * stride,
+                ),
+                values=forecast,
+                columns=series_.columns,
+                static_covariates=series_.static_covariates,
+                hierarchy=series_.hierarchy,
+            )
+        )
+    return forecasts_list if len(series) > 1 else forecasts_list[0]
+
+
+def _optimised_historical_forecasts_regression_all_points(
+    model,
+    series: Sequence[TimeSeries],
+    past_covariates: Optional[Sequence[TimeSeries]] = None,
+    future_covariates: Optional[Sequence[TimeSeries]] = None,
+    num_samples: int = 1,
+    start: Optional[Union[pd.Timestamp, float, int]] = None,
+    forecast_horizon: int = 1,
+    stride: int = 1,
+    overlap_end: bool = False,
+    show_warnings: bool = True,
+) -> Union[
+    TimeSeries, List[TimeSeries], Sequence[TimeSeries], Sequence[List[TimeSeries]]
+]:
+    """Optimized historical forecasts for RegressionModel with last_points_only = False"""
+    forecasts_list = []
+    for idx, series_ in enumerate(series):
+        past_covariates_ = past_covariates[idx] if past_covariates is not None else None
+        future_covariates_ = (
+            future_covariates[idx] if future_covariates is not None else None
+        )
+        freq = series_.freq
+
+        # obtain forecastable indexes boundaries, adjust target & covariates boundaries accordingly
+        (
+            hist_fct_start,
+            hist_fct_end,
+            hist_fct_tgt_start,
+            hist_fct_tgt_end,
+            hist_fct_pc_start,
+            hist_fct_pc_end,
+            hist_fct_fc_start,
+            hist_fct_fc_end,
+        ) = _get_historical_forecast_boundaries(
+            model=model,
+            series=series_,
+            series_idx=idx,
+            past_covariates=past_covariates_,
+            future_covariates=future_covariates_,
+            start=start,
+            forecast_horizon=forecast_horizon,
+            overlap_end=overlap_end,
+            freq=freq,
+            show_warnings=show_warnings,
+        )
+
+        # Additional shift, to account for the model output_chunk_length
+        shift_start = 0
+        shift_end = 0
+        if model.output_chunk_length > 1:
+            # used to convert the shift into the appropriate unit
+            unit = freq if series_.has_datetime_index else 1
+
+            if not model.multi_models:
+                shift_start = model.output_chunk_length - 1
+
+                hist_fct_tgt_start -= shift_start * unit
+                hist_fct_pc_start -= shift_start * unit
+                hist_fct_fc_start -= shift_start * unit
+
+            if model.output_chunk_length == forecast_horizon:
```

Comment:
can you explain why this is not required for ocl < forecast_horizon?

Reply:
it's the other way around, forecast_horizon < ocl, and it's because these timestamps will be covered by the last prediction of length `output_chunk_length`. By leaving them, the prediction will go too far.